src/vmm/src/builder.rs

```diff
@@ -763,39 +767,605 @@ pub mod tests {
+
+            assert!(vmm
+                .kernel_cmdline
+                .as_str()
+                .contains("root=PARTUUID=0eaa91a0-01 rw"));
+            assert!(vmm
+                .mmio_device_manager
+                .get_device(DeviceType::Virtio(TYPE_BLOCK), "root")
+                .is_some());
+            assert!(vmm
+                .mmio_device_manager
+                .get_device(DeviceType::Virtio(TYPE_BLOCK), "secondary")
+                .is_some());
+            assert!(vmm
+                .mmio_device_manager
+                .get_device(DeviceType::Virtio(TYPE_BLOCK), "third")
+                .is_some());
+
+            // Check if these three block devices are inserted in kernel_cmdline.
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            assert!(vmm
+                .kernel_cmdline
+                .as_str()
+                .contains("virtio_mmio.device=4K@0xd0000000:5 virtio_mmio.device=4K@0xd0001000:6 virtio_mmio.device=4K@0xd0002000:7"));
+        }
+
+        // Use case 5: root block device is rw.
+        {
+            let drive_id = String::from("root");
+            let block_configs = vec![CustomBlockConfig::new(drive_id.clone(), true, None, false)];
+            let vmm = vmm_with_block_devices(&mut event_manager, block_configs);
+            assert!(vmm.kernel_cmdline.as_str().contains("root=/dev/vda rw"));
+            assert!(vmm
+                .mmio_device_manager
+                .get_device(DeviceType::Virtio(TYPE_BLOCK), drive_id.as_str())
+                .is_some());
+        }
+
+        // Use case 6: root block device is ro, with PARTUUID.
+        {
+            let drive_id = String::from("root");
+            let block_configs = vec![CustomBlockConfig::new(
+                drive_id.clone(),
+                true,
+                Some("0eaa91a0-01".to_string()),
+                true,
+            )];
+            let vmm = vmm_with_block_devices(&mut event_manager, block_configs);
+            assert!(vmm
+                .kernel_cmdline
+                .as_str()
+                .contains("root=PARTUUID=0eaa91a0-01 ro"));
+            assert!(vmm
+                .mmio_device_manager
+                .get_device(DeviceType::Virtio(TYPE_BLOCK), drive_id.as_str())
+                .is_some());
+        }
+    }
+
+    #[test]
+    fn test_attach_vsock_device() {
+        // Placeholder for the path where a socket file will be created.
+        // The socket file will be removed when the scope ends.
+        struct TempSockFile {
+            path: String,
+        }
+
+        impl TempSockFile {
+            fn new(tmp_file: TempFile) -> Self {
+                TempSockFile {
+                    path: String::from(
+                        fs::canonicalize(tmp_file.as_path())
+                            .unwrap()
+                            .to_str()
+                            .unwrap(),
+                    ),
```

Comment:
Hmm, is the `canonicalize()` needed or can we simplify this with: `String::from(tmp_file.as_path().to_str().unwrap())`?

Reply:
I wanted to take the absolute path of the created `TempFile`. The method you propose will return the file name for a specific `TempFile`. We can use the file name, though, but it will cause problems when removing the file, at the drop time, because it will try to remove the file from the current path of test execution.

Reply:
I may be wrong, but `as_path()` doesn't return the absolute path to a file? :-?

Reply:
Actually, this is not true. `TempFile::as_path` returns a `Path` which can be relative or absolute, depending on how the `TempFile` was created, with an absolute path or a relative one. I will update as you suggested.